File: scheduler/core/src/test/java/io/orkes/conductor/scheduler/service/InMemorySchedulerDAO.java
package io.orkes.conductor.scheduler.service;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import com.netflix.conductor.common.run.SearchResult;

import io.orkes.conductor.dao.scheduler.SchedulerDAO;
import io.orkes.conductor.scheduler.model.WorkflowScheduleExecutionModel;
import io.orkes.conductor.scheduler.model.WorkflowScheduleModel;

/** Simple in-memory SchedulerDAO for OSS unit tests. Keyed by name/id. */
public class InMemorySchedulerDAO implements SchedulerDAO {

    private final Map<String, WorkflowScheduleModel> schedules = new ConcurrentHashMap<>();
    private final Map<String, WorkflowScheduleExecutionModel> executionRecords =
            new ConcurrentHashMap<>();
    private final Map<String, Long> nextRunTimes = new ConcurrentHashMap<>();

    public void clear() {
        schedules.clear();
        executionRecords.clear();
        nextRunTimes.clear();
    }

    @Override
    public void updateSchedule(WorkflowScheduleModel workflowSchedule) {
        schedules.put(workflowSchedule.getName(), workflowSchedule);
    }

    @Override
    public void saveExecutionRecord(WorkflowScheduleExecutionModel executionModel) {
        executionRecords.put(executionModel.getExecutionId(), executionModel);
    }

    @Override
    public WorkflowScheduleExecutionModel readExecutionRecord(String executionId) {
        return executionRecords.get(executionId);
    }

    @Override
    public void removeExecutionRecord(String executionId) {
        executionRecords.remove(executionId);
    }

    @Override
    public WorkflowScheduleModel findScheduleByName(String name) {
        return schedules.get(name);
    }

    @Override
    public List<WorkflowScheduleModel> findAllSchedules(String workflowName) {
        return schedules.values().stream()
                .filter(
                        s ->
                                workflowName == null
                                        || workflowName.equals(
                                                s.getStartWorkflowRequest().getName()))
                .collect(Collectors.toList());
    }

    @Override
    public void deleteWorkflowSchedule(String name) {
        schedules.remove(name);
    }

    @Override
    public List<String> getPendingExecutionRecordIds() {
        return executionRecords.values().stream()
                .map(WorkflowScheduleExecutionModel::getExecutionId)
                .collect(Collectors.toList());
    }

    @Override
    public List<WorkflowScheduleModel> getAllSchedules() {
        return new ArrayList<>(schedules.values());
    }

    @Override
    public Map<String, WorkflowScheduleModel> findAllByNames(Set<String> workflowScheduleNames) {
        Map<String, WorkflowScheduleModel> result = new HashMap<>();
        for (String name : workflowScheduleNames) {
            WorkflowScheduleModel s = schedules.get(name);
            if (s != null) {
                result.put(name, s);
            }
        }
        return result;
    }

    @Override
    public long getNextRunTimeInEpoch(String scheduleName) {
        return nextRunTimes.getOrDefault(scheduleName, -1L);
    }

    @Override
    public void setNextRunTimeInEpoch(String name, long toEpochMilli) {
        nextRunTimes.put(name, toEpochMilli);
    }

    @Override
    public SearchResult<WorkflowScheduleModel> searchSchedules(
            String workflowName,
            String scheduleName,
            Boolean paused,
            String freeText,
            int start,
            int size,
            List<String> sortOptions) {
        List<WorkflowScheduleModel> filtered =
                schedules.values().stream()
                        .filter(
                                s ->
                                        workflowName == null
                                                || workflowName.equals(
                                                        s.getStartWorkflowRequest().getName()))
                        .filter(s -> scheduleName == null || s.getName().contains(scheduleName))
                        .filter(s -> paused == null || s.isPaused() == paused)
                        .collect(Collectors.toList());

        if (sortOptions != null && !sortOptions.isEmpty()) {
            String sortOption = sortOptions.get(0);
            String[] parts = sortOption.split(":");
            String field = parts[0];
            boolean asc = parts.length < 2 || "ASC".equalsIgnoreCase(parts[1]);
            if ("name".equals(field)) {
                filtered.sort(
                        asc
                                ? Comparator.comparing(WorkflowScheduleModel::getName)
                                : Comparator.comparing(WorkflowScheduleModel::getName).reversed());
            }
        }

        int total = filtered.size();
        int end = Math.min(start + size, total);
        List<WorkflowScheduleModel> page =
                start < total ? filtered.subList(start, end) : Collections.emptyList();
        return new SearchResult<>(total, page);
    }
}

File: scheduler/core/src/test/typescript/io/orkes/conductor/scheduler/service/InMemorySchedulerDAO.ts
import { SearchResult } from '../../../../../../../../../mock';

export class InMemorySchedulerDAO {
    private schedules: Map<string, any> = new Map();
    private executionRecords: Map<string, any> = new Map();
    private nextRunTimes: Map<string, number> = new Map();

    public clear(): void {
        this.schedules.clear();
        this.executionRecords.clear();
        this.nextRunTimes.clear();
    }

    public updateSchedule(workflowSchedule: any): void {
        this.schedules.set(workflowSchedule.name, workflowSchedule);
    }

    public saveExecutionRecord(executionModel: any): void {
        this.executionRecords.set(executionModel.executionId, executionModel);
    }

    public readExecutionRecord(executionId: string): any {
        return this.executionRecords.get(executionId);
    }

    public removeExecutionRecord(executionId: string): void {
        this.executionRecords.delete(executionId);
    }

    public findScheduleByName(name: string): any {
        return this.schedules.get(name);
    }

    public findAllSchedules(workflowName: string | null): any[] {
        return Array.from(this.schedules.values())
            .filter(s => workflowName == null || workflowName === s.startWorkflowRequest?.name);
    }

    public deleteWorkflowSchedule(name: string): void {
        this.schedules.delete(name);
    }

    public getPendingExecutionRecordIds(): string[] {
        return Array.from(this.executionRecords.values()).map(r => r.executionId);
    }

    public getAllSchedules(): any[] {
        return Array.from(this.schedules.values());
    }

    public findAllByNames(workflowScheduleNames: Set<string>): Map<string, any> {
        const result = new Map<string, any>();
        for (const name of workflowScheduleNames) {
            const s = this.schedules.get(name);
            if (s !== undefined) {
                result.set(name, s);
            }
        }
        return result;
    }

    public getNextRunTimeInEpoch(scheduleName: string): number {
        return this.nextRunTimes.get(scheduleName) ?? -1;
    }

    public setNextRunTimeInEpoch(name: string, toEpochMilli: number): void {
        this.nextRunTimes.set(name, toEpochMilli);
    }

    public searchSchedules(
        workflowName: string | null,
        scheduleName: string | null,
        paused: boolean | null,
        freeText: string | null,
        start: number,
        size: number,
        sortOptions: string[] | null
    ): any {
        let filtered = Array.from(this.schedules.values())
            .filter(s => workflowName == null || workflowName === s.startWorkflowRequest?.name)
            .filter(s => scheduleName == null || s.name.includes(scheduleName))
            .filter(s => paused == null || s.paused === paused);

        if (sortOptions && sortOptions.length > 0) {
            const sortOption = sortOptions[0];
            const parts = sortOption.split(':');
            const field = parts[0];
            const asc = parts.length < 2 || parts[1].toUpperCase() === 'ASC';
            if (field === 'name') {
                filtered.sort((a, b) => {
                    const cmp = a.name.localeCompare(b.name);
                    return asc ? cmp : -cmp;
                });
            }
        }

        const total = filtered.length;
        const end = Math.min(start + size, total);
        const page = start < total ? filtered.slice(start, end) : [];
        return { totalHits: total, results: page };
    }
}
